onnxruntime/contrib_ops/rocm/bert/attention.cu

```diff
@@ -40,7 +40,11 @@ REGISTER_KERNEL_TYPED(MLFloat16)
 
 template <typename T>
 Attention<T>::Attention(const OpKernelInfo& info)
-    : RocmKernel(info), AttentionBase(info, true), attn_type_(kAttention) {}
+    : RocmKernel(info), AttentionBase(info, true), attn_type_(kAttention) {
+  using HipT = typename ToHipType<T>::MappedType;
+  using AttentionTunableOp = GemmSoftmaxGemmPermuteTunableOp<HipT>;
+  tunable_op_ = std::make_shared<AttentionTunableOp>();
```

Comment:
> std::make_shared<AttentionTunableOp>();

As far as I can tell, this instance isn't shared between different instances of `Attention`. So, I have two questions:

1. Would it make sense for different `Attention` objects to share a single `AttentionTunableOp`? If so, should we consider making `AttentionTunableOp` a singleton?
2. If it does not make sense for different `Attention` objects to share a single `AttentionTunableOp`, then should this be a `unique_ptr`, rather than a `shared_ptr`?

Reply:
No and No, because opt for `shared_ptr` is merely to take advantage of its type erasure, not the ownership. To use `unique_ptr` ,then you still need to encode the type in the deleter (manually), then include the `.cuh` in a `.h`...